projects/components/src/download-file/service/file-download.service.ts

```diff
@@ -31,20 +31,28 @@ export class FileDownloadService {
    * @param config Csv download config
    */
   public downloadAsCsv(config: CsvDownloadFileConfig): Observable<FileDownloadEvent> {
+    const getAllDataKeys = (data: Dictionary<unknown>[]): string[] => uniq(data.map(row => Object.keys(row)).flat());
     // If given header is empty, then create header from the data keys
     const header$ = isEmpty(config.header)
       ? config.dataSource.pipe(
-          map(data => Object.keys(data[0] ?? [])),
-          map(keys => keys.map(startCase))
+          map(
+            data => getAllDataKeys(data),
+            map((keys: string[]) => keys.map(startCase))
+          )
         )
       : of(config.header!);
 
     // Value replacer for null and undefined values
     const replacer = (_: string, value: string) => value ?? '-';
 
     // Convert values into strings
-    const values$ = config.dataSource.pipe(
-      map(data => data.map(datum => Object.values(datum).map(value => JSON.stringify(value, replacer))))
+    const values$: Observable<string[][]> = config.dataSource.pipe(
+      // All rows are not guaranteed to have all keys, so add all keys to all rows
+      map(data => {
+        const allKeys = getAllDataKeys(data);
```

Comment:
nit: now that I read this a little closer, we're being pretty inefficient on doing this. We're iterating all the values to calculate the keys twice, iterating to calculate the values once and then iterating again to merge them. If you think that makes it more readable, by all means go for it. Personally, I think it does the opposite - I'd try to pipe off the data source a single time.

More importantly though, I see the header can be specified via config. If that's the case, we still have an issue. If I give a config of "header1, header2" and data containing extra keys (or keys in a different order) both the prior code and our new code would produce mismatched rows. However we calculate the header, we need to use that same result to extract the data.

Reply:
Updated: [`1a62ad4` (#2478)](https://github.com/hypertrace/hypertrace-ui/pull/2478/commits/1a62ad49e92d0f63a75450ee3b546bfd7af2052e)